src/secure_random.cr

```diff
@@ -58,14 +74,80 @@ module SecureRandom
       raise ArgumentError.new "negative size: #{n}"
     end
 
-    slice = Slice(UInt8).new(n)
-    result = LibCrypto.rand_bytes slice, n
-    if result != 1
-      error = LibCrypto.err_get_error
-      error_string = String.new LibCrypto.err_error_string(error, nil)
-      raise error_string
+    init unless @@initialized
+
+    ifdef linux
+      if @@getrandom_available
+        return getrandom(n)
+      end
+    end
+
+    buf = Slice(UInt8).new(n)
+
+    if urandom = @@urandom
+      urandom.read_fully(buf)
+      return buf
+    end
+
+    ifdef !without_openssl
+      ret = LibCrypto.rand_bytes(buf, n)
+      raise String.new(LibCrypto.err_error_string(LibCrypto.err_get_error, nil)) unless ret == 1
+      return buf
+    end
+
+    raise "Failed to access secure source to generate random bytes!"
+  end
+
+  private def self.init
+    @@initialized = true
+
+    ifdef linux
+      if getrandom(Slice(UInt8).new(16)) >= 0
+        @@getrandom_available = true
+        return
+      end
+    end
+
+    @@urandom = urandom = File.open("/dev/urandom", "r")
+    urandom.sync = true # don't buffer bytes
+  end
+
+  ifdef linux
+    @@getrandom_available = false
+
+    # Reads n random bytes using the Linux `getrandom(2)` syscall.
+    private def self.getrandom(n : Int)
+      Slice(UInt8).new(n).tap do |buf|
+        # getrandom(2) may only read up to 256 bytes at once without being
+        # interrupted or returning early
+        chunk_size = 256
+
+        while buf.size > 0
+          if buf.size < chunk_size
+            chunk_size = buf.size
+          end
+
+          read_bytes = getrandom(buf[0, chunk_size])
+          raise Errno.new("getrandom") if read_bytes == -1
+
+          buf += read_bytes
+        end
+      end
+    end
+
+    # Low-level wrapper for the `getrandom(2)` syscall, returns the number of
+    # bytes read or `-1` if an error occured (or the syscall isn't available)
+    # and sets `Errno.value`.
```

Comment:
I think explaining here why `syscall(2)` is used over binding `getrandom(2)` would be good :)